### ISSUE TITLE
Could the CF search also search the canonical unit field?

### ISSUE THREAD

**Issue by larsbarring** [cf-convention/cf-convention.github.io#200]
_The functionality discussed in this issue has been implemented by @bzah in_ [this pull request](https://github.com/cf-convention/cf-convention.github.io/pull/218).

In relation to the CF [discussion issue  #101](https://github.com/cf-convention/discuss/issues/101) it would be useful to search for standard names having  a specific canonical unit. Would it be reasonably simple to extend the standard name search functionality to enable this? 

A related request was raised in [issue #59](https://github.com/cf-convention/cf-convention.github.io/issues/59).

**Comment by JonathanGregory** [comment]
@bzah offered to investigate possibilities for doing this. It's being discussed in [issue 59](https://github.com/cf-convention/cf-convention.github.io/issues/59).

**Comment by bzah** [comment]
Unit search is implemented in #218 as long as the alias search (#50).
See the demonstrator:  https://bzah.github.io/cf-convention.github.io/Data/cf-standard-names/current/build/cf-standard-name-table.html
I tried only simple use cases.

I usually prefer to implement different things in different PR but in the case of CF website it could be fine. If it's not, I don't mind splitting those changes into two pull requests.

**Comment by bzah** [comment]
Note that the AND and OR features are ignored in units search as the space character can be part of a unit (e.g. "m-1 s").

**Comment by JonathanGregory** [comment]
Dear @bzah 

Thanks a lot for working on this. If I understand correctly, this is a text search in the units string as well as the standard name string, is that right? Thus, for example, if I specify `k` as the search string, it finds standard names which contain `k` or whose units contain `k` or `K`. Certainly that can find entries with a particular unit, but I wonder whether @larsbarring had in mind something more specific. For example, you could have a second search string (another text entry field on the page) to specify the canonical unit exactly and case-sensitively. Entering `K` in this field would find all standard names whose canonical unit is exactly `K`, but not `kg` or `K s-1`, for example. If you enter both search strings, they would be ANDed, so you can find standard names which contain a certain string and have a certain unit. Is that something like what you were thinking, Lars?

Best wishes

Jonathan

**Comment by bzah** [comment]
Hi @JonathanGregory 
would something like
```
Search
[_____________]
⚫ AND  O OR (warning, binary keyword is ignored in unit search)
Fields:
[x] standard_name  [  ] help text
[x] aliases        [  ] units
```
where user can enable/disable each individual criteria (including the current default standard_name),
works for you ?
(hopefully my "ascii art" is understandable!)

**Comment by JonathanGregory** [comment]
Dear Abel @bzah

I like ASCII art, thanks. I agree that this would work well for the standard name, aliases and help text. It's good to have that flexibility. Still it seems to me that the units require different treatment, for two reasons

* The units are canonical, and there are rather few distinct possibilities. I think, therefore, that for units you probably want an exact match with the entire string. I think it's likely that if you specify "kg", for instance, you don't want "kg m-2".

* Whereas standard names are all lower case and could be searched case-insensitively, units are mixed case and case-sensitive. For example, `s` and `S` are different units.

Consequently, I think that two search strings are needed, differently treated, one for standard names, aliases and help text, the other for units. Either could be left blank or both could be specified. But it would be usefult to know what @larsbarring and others think. Maybe my idea of this is more complicated than needed by users, and too hard to implement.

Best wishes

Jonathan

**Comment by bzah** [comment]
Hi Jonothan, I see your point.
In such case: exact matches + not too many values + all possible values are known, I would prefer to use a drop-down menu. So that users don't need to bother writing the exact CF unit wording (e.g. m s-1 ; m/s ; m*s-1...).

By default the drop-down menu would have a "any units" value, then the user would be able to choose a specific units and combine it with the search bar query when pressing filter button.

**Comment by JonathanGregory** [comment]
Dear Abel

That's a good idea, and I agree that it avoids the need to get the formatting exactly right. But there are more than I guessed when I said "rather few". There are about 100 distinct units - a lot fewer than the number of standard names, of course, but perhaps too many for a drop-down menu?

Best wishes

Jonathan

**Comment by larsbarring** [comment]
When we now are trying to resolve the [discuss/#101](https://github.com/cf-convention/discuss/issues/101#issuecomment-1519009630) issue it would be very helpful if pull request #218 could be pushed over the line and implemented. 

Regarding the previous 2-3 comments in this issue, I imagine that implementing a (likely way too long) dropdown menu, or some kind of unit parsing and harmonising of unit strings involves some substantial design and development work. Thus I think that we should not let Perfect be the enemy of the Good. So I suggest that what is now (or soon after final polishing) is in the PR can be put into use. Then we gain experience that may (or may not) lead to further suggestions for improvements.

What do you think @JonathanGregory @bzah @sadielbartholomew @feggleton @japamment ?

**Comment by bzah** [comment]
Hi Lars, I agree with you, I would also like to see this move forward. Note that the demonstrator is still online if needed for https://github.com/cf-convention/discuss/issues/101, see: https://bzah.github.io/cf-convention.github.io/Data/cf-standard-names/current/build/cf-standard-name-table.html

As for the code review, the git diff is huge only because:
- I had to re-generate the html output (~10K lines)
- I included the python script I used to generate the html page from the xsl
- I also reformatted the JS script to JS standards (if I remember correctly I used [Prettier](https://github.com/prettier/prettier)).

If any of these changes should not be part of this PR, I can remove them.

**Comment by larsbarring** [comment]
Hi Abel,

aah ... and now when again testing the demonstrator, I recall one possible improvement that I though of last time (and I think is what @JonathanGregory was aiming at in his [comment](https://github.com/cf-convention/cf-convention.github.io/issues/200#issuecomment-1364505605):

Would it be easy (!!) to adjust the search functionality so that when specifying search in the`units` field it the only field included?  I.e. something like this (building on you ascii art):
``` txt
.... .... ....
[ ] Also search help text
[ ] Also search aliases text
[ ] Only search units text
```
If not all that easy, the I will be happy with what is already present in the demonstrator :+1:

**Comment by bzah** [comment]
It doesn't look too difficult, I will try to make the changes and let you know.

**Comment by larsbarring** [comment]
... and, just to clarify what I meant: the search is only in the units text, but return results for the whole entry/-ies just as for a "normal search".

**Comment by JonathanGregory** [comment]
I agree with @larsbarring that it's useful to search _only_ on the `units` (but return the search as normal), and (as before) I think that the search on the `units` should be an exact match and case-sensitive. In the interests of progress, as Lars says, we can live with the problem of units being written in various equivalent ways. The facility will be useful anyway. Thanks a lot, @bzah.

**Comment by larsbarring** [comment]
Hi Abel  @bzah --- I have made some quick test and it indeed meets what I had in mind when first asking for this. This will be very useful!! Many thanks /Lars

**Comment by larsbarring** [comment]
Dear Jonathan @JonathanGregory 
I both agree and don't, i.e. I can see good use cases for both alternatives.
* Search for partial string: like`C`, to find all occurrences of units that includes a degree_C (including any allowed aliases), and similar for `K`. Then I can  accept to get some irrelevant answers while knowing that *all* occurrences where found. And I guess that the [error](https://github.com/cf-convention/discuss/issues/230) that you incidentally found was not by searching for the full unit `J Kg-1`.
* Searching for a full string (exact match): if I want to know all standard names involving e.g. `K s-1`, or `K m-1`, or whatever I might be interested in.

**Comment by bzah** [comment]
Ah I miss that part about the entire string matching the query.
Fortunately, it is already implemented as the query accept a regular expression. So, if you want to match exactly `kg m-3`, you can query `^kg m-3$` instead!

I found out regular expression were allowed while working on the code, I did not specifically implemented it myself. 
Regexes are enabled on any search as long as there is no space character (`" "`) in the query, in which case the boolean AND|OR criteria is enabled and regexes are disabled.

Knowing, this, do you people still want a specific handling for exact matches ? 
In any case, I can add a note about regexes and exact matches with it on the page.

**Comment by JonathanGregory** [comment]
I'm very happy with regexp, because that gives a lot of flexibility! Not everyone is familiar with regexp, however, so it would be helpful to write the syntax explicitly on the web page. Possibly you could implement something slightly easier, which you could substitute in your program? For example, `=kg m-2` might be used to indicate an match to the entire `units` string, which you could translate to `^kg m-3$` before using it. I think then it would be as easy to use as a separate checkbox, for example.

I don't understand quite what you mean by _Regexes are enabled on any search as long as there is no space character (" ") in the query_. There's a space in `^kg m-3$`, but it's interpreted as a regexp.

Thanks for working on this.

**Comment by bzah** [comment]
I think a collapsible section to explain these details would be nice.
I will also try to add the prefix `=` as the mean to recognize an exact match,

**Comment by davidhassell** [comment]
Hello,

I support everything here, except that I'm not so keen on Jonathan's suggestion of a `=` new regex syntax. I think that for those who do know regex syntax, a new way of doing it confusing; and I think `^...$` is reasonably easy for people to pick up, given that there would be some guidance on hand.

Thanks,
David

**Comment by bzah** [comment]
Please have a look at the help text I have added : [demonstrator ](https://bzah.github.io/cf-convention.github.io/Data/cf-standard-names/current/build/cf-standard-name-table.html)
Note that English is not my first language so feel free to propose any rewording!

**Comment by davidhassell** [comment]
Hello @bzah,

I've had a play with your [demonstrator](https://bzah.github.io/cf-convention.github.io/Data/cf-standard-names/current/build/cf-standard-name-table.html), and I think that it's great.

I have two suggestions:

* Change the "Only search units" text by the radio button and in the advanced text to "Only search canonical units".
* It took me a while to realize that the guidance I knew existed was buried inside the "Advanced searches" collapsed section, as I didn't think I was doing an advanced search! Can this text always be displayed?

Thanks,
David

**Comment by bzah** [comment]
Hi Jonathan, David. I'm just waiting for the potential feedback of others and will included theses changes as a bulk.

> It took me a while to realize that the guidance I knew existed was buried inside the "Advanced searches" collapsed section, as I didn't think I was doing an advanced search! Can this text always be displayed?
Yes.

**Comment by larsbarring** [comment]
Hello Abel @bzah 

I was just to write the same as David: It is indeed a great tool that already has proven its worth as several problematic canonical units have been discovered in the testing phase. And I, too, that the "Advanced searches" hides useful information; either display it all the time, or write something like "Click here for more detailed help on search options".

Cheers
Lars

**Comment by bzah** [comment]
The issue is that the "Advanced search" label doesn't look clickable indeed and, to add the styles that would make it look like it, we would need to have a proper CSS style sheet, which goes beyond the scope of this PR.

Anyway, I made the section unrolled by default and if the user is annoyed by it it can still be collapsed.
I'm happy to include any other change you deem necessary!

**Comment by larsbarring** [comment]
I have just had a look at the [Demonstrator](https://bzah.github.io/cf-convention.github.io/Data/cf-standard-names/current/build/cf-standard-name-table.html) after the last round of updates (in PR #218).  I think it looks great!

If (when) everyone else is happy, and I guess we are nearly there, what else is needed to put this into operation?
* A code review of the PR  --- perhaps by @sadielbartholomew ?
* Is something more needed to integrate this with the procedure for publishing new standard name tables?
* Merging into `main`

**Comment by davidhassell** [comment]
I agree - it's really good. Thanks, Abel @bzah, for all of work on this.

**Comment by JonathanGregory** [comment]
Yes, me too, that's very useful indeed. Jonathan

**Comment by japamment** [comment]
Hi @bzah 

As requested, I have reviewed pull request [#218](https://github.com/cf-convention/cf-convention.github.io/pull/218) and have made a comment in the discussion there.

I've tried the Demonstrator and am happy with the results it is producing. Thank you for working on this - I agree that the alias and units searches are very useful additions to the standard name table and will be widely used. I'd be very happy to see these changes implemented soon.

Once the pull request has been accepted and merged I will reprocess the current and previous versions of the standard name table to make the new searches visible.

Best wishes,
Alison
